skbio/metadata/_interval.py

```diff
@@ -726,6 +741,44 @@ def __repr__(self):
 
         return '\n'.join([l1, l2] + items)
 
+    @experimental(as_of='0.5.0-dev')
+    def __copy__(self):
+        '''Return a shallow copy.
+
+        Notes
+        -----
+        The references to original ``Interval`` objects are created
+        in the returned copy object.
+
+        See Also
+        --------
+        deepcopy
+        '''
+        return self._copy(False, {})
+
+    @experimental(as_of='0.5.0-dev')
+    def __deepcopy__(self, memo):
+        '''Return a deep copy.
+
+        See Also
+        --------
+        copy
+        '''
+        return self._copy(True, memo)
+
+    @experimental(as_of='0.5.0-dev')
+    def _copy(self, deep, memo):
```

Comment:
Copying here is a little tricky. If we copy the `Interval` objects in this `IntervalMetadata`, they will still refer to this object as their parent. Any subsequent updates to these copied `Interval` objects will only affect *this* `IntervalMetadata` object, while we'd want those updates to be reflected in both the current object and the copied object (e.g. staled tree state). I don't think this behavior is possible with the current design because each `Interval` has only a single parent.

Since `IntervalMetadata` and `Interval` are inherently linked (bidirectional), we need to deep copy both types of objects. The only part of `Interval` that isn't necessarily deep-copied is the `.metadata` dict, since that can be safely shallow- or deep-copied using `copy.copy` or `copy.deepcopy`, respectively.

Here's one way to do this:

```python
cp = IntervalMetadata(self.upper_bound)

for interval in self._intervals:
    # Only need to shallow-copy `locations` and `boundaries` because their elements are immutable.
    locations_cp = interval.locations[:]
    boundaries_cp = interval.boundaries[:]
    if deep:
        metadata_cp = copy.deepcopy(interval.metadata, memo)
    else:
        metadata_cp = copy.copy(interval.metadata)
    
    cp.add(locations_cp, boundaries=boundaries_cp, metadata=metadata_cp)

return cp
```

Does this copying strategy make sense?

Once we've finalized the copying behavior, can you also update the unit tests so that they would have failed with the current implementation?